Move email sender options to the email template editor

lysyjan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes #56123.

- This PR moves the email sender options from `WooCommerce -> Settings -> Emails` to the new email template editor, which is behind a feature flag `block_email_editor`.
- When the feature flag `block_email_editor` is active, a user can see a notice in the `WooCommerce -> Settings -> Emails`.

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

| Before | After |
| ------ | ----- |
|<img width="1278" alt="Screenshot 2025-03-17 at 20 24 30" src="https://github.com/user-attachments/assets/6be369ac-160e-405d-be38-105f044a9ff0" /> | <img width="1279" alt="Screenshot 2025-03-17 at 20 30 34" src="https://github.com/user-attachments/assets/8f06a729-7a03-4261-9cfa-a558607cf1e2" />|  
|<img width="1280" alt="Screenshot 2025-03-17 at 20 31 33" src="https://github.com/user-attachments/assets/58863da5-eca4-421c-916c-1f7268411631" /> | <img width="1279" alt="Screenshot 2025-03-17 at 20 31 00" src="https://github.com/user-attachments/assets/15a9228e-6596-4178-a81f-dc167b07ca29" /> |
|<img width="1277" alt="Screenshot 2025-03-17 at 20 31 50" src="https://github.com/user-attachments/assets/18c44453-2ce1-47ed-b633-30368b997f61" />|<img width="1282" alt="Screenshot 2025-03-17 at 20 38 38" src="https://github.com/user-attachments/assets/77dbdc6d-9902-4292-bd1b-cbb0f674f524" />|
||<img width="492" alt="Screenshot 2025-03-19 at 10 28 17" src="https://github.com/user-attachments/assets/ff298022-2ba0-411e-a3cd-29a5f5a0e961" />|
||<img width="1582" alt="Screenshot 2025-03-19 at 10 28 36" src="https://github.com/user-attachments/assets/b45733bb-a8d1-43f0-b9f3-a7629f612d23" />|



<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Check that saving the sender name and email address still works in `WooCommerce -> Settings -> Emails` when the feature `block_email_editor` is disabled.
2. Enable the feature flag `block_email_editor`
  a. A notice about moving the sender options is visible only in `Settings` and only for the `Emails` tab
  b. The notice is closable
  c. The notice is not visible anymore when the feature flag `block_email_editor` is deactivated
3. The sender options can be set in the email template editor after opening the sidebar related to the template
  a. When an invalid email address is typed, the input should display an HTML validation warning.
  b. Saving an invalid email address is not allowed, and a notice about an invalid email is displayed after clicking on save.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [x] This Pull Request does not require a changelog entry. (Comment required below)

Because this feature is still behind the feature flag I think we don't need changelog here.
<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

lysyjan committed:
Add TemplateSettingsPanel component to email editor sidebar

- Introduced a new TemplateSettingsPanel component that allows plugins to add custom template sections.
- Integrated TemplateSettingsPanel into the existing TemplateSettings component for enhanced functionality.
#56123

lysyjan committed:
Add TemplateSenderPanel and modify email editor sidebar

- Introduced TemplateSenderPanel component to enhance the email editor sidebar with sender options.
- Implemented modifyTemplateSidebar function to add a custom section in the email editor.
- Added corresponding styles for the new template settings section.

This update improves the customization capabilities of the email editor for WooCommerce users.

lysyjan committed:
Add style registration for email editor integration

#56123

lysyjan committed:
Enhance email editor with sender settings integration

- Added sender settings to the email editor, allowing customization of "from" name and email address.
- Implemented TemplateApiController to manage retrieval and storage of sender settings.
- Updated TemplateSenderPanel to utilize new sender settings functionality.

lysyjan committed:
Add conditional sender options display

#56123

lysyjan committed:
Add email sender options notice

- Introduced a new admin notice for email sender options, informing users about the relocation of these settings to the email template editor.
- Implemented conditional display logic for the notice based on the block editor feature status.
- Added a corresponding view file for the notice.

This update enhances user awareness regarding email settings management in WooCommerce. #56123

lysyjan committed:
Add unit tests for TemplateApiController

- Introduced tests for the TemplateApiController class to validate the retrieval and saving of WooCommerce email template data.
- Included tests for both WooCommerce and non-WooCommerce templates to ensure correct behavior in different scenarios.

#56123

lysyjan committed:
Enhance SaveAllButton functionality with content validation

- Updated SaveAllButton component to accept a validateContent prop, ensuring content is validated before saving.
- Modified Header component to pass validateContent to SaveAllButton when applicable.

This change improves the reliability of the save operation in the email editor.
#56123

lysyjan committed:
Add email validation rule and enhance sender input handling

- Introduced an email validation rule in the email editor to ensure a valid sender email address is provided.
- Updated TemplateSenderPanel to include HTML5 validation for the "from" email input, improving user feedback on input errors.

This enhancement increases the reliability of email settings in the WooCommerce email editor.

lysyjan committed:
Add email validation in TemplateApiController

- Implemented validation for the sender email address in the save_template_data method, throwing an InvalidArgumentException for invalid email formats.
- Added a unit test to ensure the exception is correctly thrown when an invalid email address is provided.

#56123

lysyjan committed:
Refactor email validation in email editor integration

- Updated the email validation logic in the email editor to use HTML5 input validation, enhancing user experience by providing immediate feedback on invalid email formats.
- Modified the TemplateApiController to align with the new validation approach, ensuring consistency across the email settings management.

#56123

costasovo commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/email-editor-integration/templates/template_sender_panel.tsx:
```diff
@@ -0,0 +1,49 @@
+				<PanelRow>
+					<TextControl
+						className="woocommerce-email-sidebar-template-settings-sender-options-input"
+						label={ __( '“from” name', 'woocommerce' ) }
```
This translation would perhaps use a comment for translators. Maybe something like:
```
/* translators: Label for the sender's name in email settings. */
```
I would add it also for “from” email

costasovo commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/email-editor-integration/templates/style.scss:
```diff
@@ -0,0 +1,12 @@
+		margin-top: 0;
+	}
+	p {
+		color: #757575;
```
I think you can also use `$gray-700` variable for that color.

costasovo commented on plugins/woocommerce/src/Internal/EmailEditor/EmailTemplates/TemplateApiController.php:
```diff
@@ -0,0 +1,67 @@
+	 * @return array
+	 */
+	public function get_template_data_schema(): array {
+		return Builder::object(
+			array(
+				'sender_settings' => Builder::object(
+					array(
+						'preheader'   => Builder::string(),
+						'preview_url' => Builder::string(),
+					)
+				),
+			)
+		)->to_array();
```
Oh! Nice. I didn't realize we brought the builder with the package. 👍

costasovo commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/email-editor-integration/templates/template_sender_panel.tsx:
```diff
@@ -3,8 +3,32 @@
+				from_address: value,
+			},
+		} );
+	};
```
We could add useCallback to avoid redefining the functions on each re-render.

costasovo commented on plugins/woocommerce/src/Internal/EmailEditor/EmailTemplates/TemplateApiController.php:
```diff
@@ -0,0 +1,67 @@
+		if ( WooEmailTemplate::TEMPLATE_SLUG === $template_post->slug && isset( $data['sender_settings'] ) ) {
+			update_option( 'woocommerce_email_from_name', $data['sender_settings']['from_name'] );
+			update_option( 'woocommerce_email_from_address', $data['sender_settings']['from_address'] );
+		}
```
Just out of curiosity. I see we don't check the `from_name` key is present in `$data['sender_settings']['from_name']`. Does the API do this via the schema? I assume it does, but I'm not sure.

lysyjan replied:
This is good point. The schema does not check it because the endpoint should update data only for the specific template only. I will add a code with some checks.

costasovo commented on plugins/woocommerce/includes/admin/settings/class-wc-settings-emails.php:
```diff
@@ -277,6 +277,8 @@ protected function get_settings_for_default_section() {
 		$body_text_color_setting_in_palette   = $reorder_colors ? $body_text_color_setting : null;
 		$footer_text_color_setting_in_palette = $reorder_colors ? $footer_text_color_setting : null;
 
+		$block_email_editor_enabled = get_option( 'woocommerce_feature_block_email_editor_enabled', 'no' );
```
I think we should rather use  `FeaturesUtil::feature_is_enabled( 'block_email_editor' )` for this check, and we get `true` or `false`. I think it is better than checking the `no` string as it may theoretically be '0' or something else.

lysyjan replied:
Thanks for this. I had it in my mind, but in the end, I forgot.

costasovo reviewed (commented):
@lysyjan Great job! I left a few comments, but they are just tiny details. Thanks for adding the unit test. There is also one question or possible issue I run into during testing. Please see below:

### Testing performed

✅ When the feature flag is disabled saving of sender's name and address works as expected
✅ After enabling the flag, I can see the notice and the sender's name and inputs are no longer displayed. Note: I noticed that the link Email Template goes to the email listing. I think that's fine for now because as for now we can't link the template mode directly.
❓  I noticed that when I dismiss the notice and switch between tabs in the settings the notice pops up again. I would expect it to stay hidden. Is that expected?
✅ I tested saving the sender's email and name settings in the editor and it works as expected.
✅ I tested adding an invalid email and validation worked as expected.
✅ After disabling the feature flag the notice is not displayed any more.

**Video:**
I started with a feature flag `off` then set it 'on` and towards the end I set it again `off` to verify that the notice is gone. I didn't record the validation testing so I'm providing a screenshot.

https://github.com/user-attachments/assets/32d32175-42fc-4f17-8912-267b5c00c116

**Screenshot of validation message:**

<img width="1342" alt="Screenshot 2025-03-20 at 11 46 53" src="https://github.com/user-attachments/assets/fc201ceb-cae8-43f9-8a5b-9c821d2ac08f" />

I also noticed a warning about the script size, but I consider that fine as it is a new script, and the test obviously checks the percentage. Overall size IMO grew accordingly.

![Screenshot 2025-03-20 at 11 53 38](https://github.com/user-attachments/assets/c8e6bd2a-efcc-4170-8346-5a21c88b33f7)

lysyjan committed:
Fix user meta key name for dismissed email sender options notice

#56123

lysyjan committed:
Improve accessibility of sender options labels in TemplateSenderPanel

- Updated label attributes for the "from" name and "from" email fields to include translator comments, enhancing clarity for translators and improving accessibility for users.

#56123

lysyjan committed:
Update email editor styles to use SCSS variable for text color

- Changed the paragraph text color in the email editor styles from a hex value to a SCSS variable ($gray-700), improving maintainability and consistency in styling.

#56123

lysyjan committed:
Refactor email editor feature flag handling

- Updated the method for checking the status of the block email editor feature, replacing the direct option retrieval with a utility function for improved clarity and maintainability.
- Adjusted the conditional logic to enhance readability.

lysyjan committed:
Enhance email sender settings update logic in TemplateApiController

- Improved the save_template_data method to conditionally update the sender name and email address only if they have changed, reducing unnecessary database updates.
- Added null checks for sender settings to ensure valid data is processed, enhancing the robustness of the email settings management.

#56123

lysyjan committed:
Refactor TemplateSenderPanel to use useCallback for state update functions

- Updated handleFromNameChange and handleFromAddressChange to utilize useCallback, improving performance by memoizing the functions and preventing unnecessary re-renders.

#56123

costasovo commented on plugins/woocommerce/includes/admin/class-wc-admin-notices.php:
```diff
@@ -673,7 +673,7 @@ public static function email_sender_options_notice() {
 			$notice_dismissed = apply_filters(
 				'woocommerce_hide_email_sender_options_notice',
-				get_user_meta( get_current_user_id(), 'dismissed_block_email_editor_notice', true )
+				get_user_meta( get_current_user_id(), 'dismissed_email_sender_options_notice', true )
```
I missed this in the first review. I see that in the newer code in the [src](https://github.com/search?q=repo%3Awoocommerce%2Fwoocommerce+get_user_meta+path%3A%2F%5Eplugins%5C%2Fwoocommerce%5C%2Fsrc%5C%2F%2F+&type=code) the user meta is often prefixed `woocommerce_`. Perhaps we could do that as well so that it is less conflict-prone.

costasovo reviewed (changes requested):
@lysyjan Thanks for the update.

I added one comment with a suggestion, but when I started testing I noticed a visual issue in the settings > email page.

When the I activate the feature flag, I see that the preview is partially over the emails table.
It is possible that the issue was present in the first testing I did, but I had a smaller window so I missed it.

I found this was caused by commit [87235c0b01cccaf6e98fc6c4f5f5eaa6a8cb65a6](https://github.com/woocommerce/woocommerce/pull/56473/commits/87235c0b01cccaf6e98fc6c4f5f5eaa6a8cb65a6)

![Screenshot 2025-03-21 at 11 22 03](https://github.com/user-attachments/assets/f42996c1-b3e2-4be1-a368-b1f7237c4f9e)

I think we need to address the issue.

lysyjan committed:
Change name of the email sender option notice

#56123

lysyjan commented:
@costasovo I added a new commit renaming the option. I also rebased the branch and cannot replicate the issue on the email settings page anymore.

costasovo reviewed (approved):
@lysyjan Thank you for updating the name of the option!

I wasn't able to replicate the issue with the preview. So it was probably only a glitch on my site.

### Testing performed

Video of basic testing:

https://github.com/user-attachments/assets/97f7526c-fc02-4091-9eab-f3cbe54405b7

Screenshot of received email with updated sender values:
<img width="1257" alt="Screenshot 2025-03-30 at 16 14 47" src="https://github.com/user-attachments/assets/3a9ec03c-3bcc-43bb-b59f-ac20214b8056" />
